jax/api.py

```diff
@@ -588,27 +588,40 @@ def hessian(fun: Callable, argnums: Union[int, Sequence[int]] = 0,
-          'b': {'a': DeviceArray([[[ 1.      ,  0.      ], [ 0.      ,  0.      ]],
-                                  [[ 0.      ,  0.      ], [ 0.      , 12.317766]]], dtype=float32),
-                'b': DeviceArray([[[0.      , 0.      ], [0.      , 0.      ]],
-                                 [[0.      , 0.      ], [0.      , 3.843624]]], dtype=float32)}}}
-
-  If the usual dense matrix form of the Hessian is desired, one can
-  flatten and concatenate the arguments into a single 1D array before
-  computing the Hessian.
+  :py:func:`hessian` is a generalization of the usual definition of the Hessian
+  that supports nested Python containers (i.e. pytrees) as inputs and outputs.
+  The tree structure of ``jax.hessian(fun)(x)`` is given by forming a tree
+  product of the structure of ``fun(x)`` with a tree product of two copies of
+  the structure of ``x``. A tree product of two tree structures is formed by
+  replacing each leaf of the first tree with a copy of the second. For example:
+
+  >>> f = lambda dct: {"c": jnp.power(dct["a"], dct["b"])}
+  >>> print(jax.hessian(f)({"a": jnp.arange(2.) + 1., "b": jnp.arange(2.) + 2.}))
+  {'c': {'a': {'a': DeviceArray([[[ 2.,  0.], [ 0.,  0.]],
+                                 [[ 0.,  0.], [ 0., 12.]]], dtype=float32),
+               'b': DeviceArray([[[ 1.      ,  0.      ], [ 0.      ,  0.      ]],
+                                 [[ 0.      ,  0.      ], [ 0.      , 12.317766]]], dtype=float32)},
+         'b': {'a': DeviceArray([[[ 1.      ,  0.      ], [ 0.      ,  0.      ]],
+                                 [[ 0.      ,  0.      ], [ 0.      , 12.317766]]], dtype=float32),
+               'b': DeviceArray([[[0.      , 0.      ], [0.      , 0.      ]],
+                                [[0.      , 0.      ], [0.      , 3.843624]]], dtype=float32)}}}
+
+  Thus each leaf in the tree structure of ``jax.hessian(fun)(x)`` corresponds to
+  a leaf of ``fun(x)`` and a pair of leaves of ``x``. For each leaf in
+  ``jax.hessian(fun)(x)``, if the corresponding array leaf of ``fun(x)`` has
+  shape ``(out_1, out_2, ...)`` and the corresponding array leaves of ``x`` have
+  shape ``(in_1_1, in_1_2, ...)`` and ``(in_2_1, in_2_2, ...)`` respectively,
+  then the Hessian leaf has shape ``(out_1, out_2, ..., in_1_1, in_1_2, ...,
+  in_2_1, in_2_2, ...)``. In other words, the Python tree structure represents
+  the block structure of the Hessian, with blocks determined by the input and
+  output pytrees.
+
+  In particular, an array is produced (with no pytrees involved) when the
+  function input ``x`` and output ``fun(x)`` are each a single array, as in the
+  ``g`` example above. If ``fun(x)`` has shape ``(out1, out2, ...)`` and ``x``
+  has shape ``(in1, in2, ...)`` then ``jax.hessian(fun)(x)`` has shape
+  ``(out1, out2, ..., in1, in2, ..., in1, in2, ...)``. To flatten pytrees into
+  1D vectors, consider using ``jax.flatten_util.flatten_pytree``.
```

Comment:
This isn't in the public (documented) API. Do we want it to be?

Reply:
Yeah seems reasonable to include. I think it might have been added specifically for this use case actually. (Someone wanted to store parameters in pytrees, but then examine the spectrum of the Hessian and so didn't care to have the pytree structure preserved in the Hessian.)